editor: Fix hang when scrolling over single line input fields

MrSubidubi opened the pull request:
Closes #21684
Closes #28463
Closes #28264 

This PR fixes Zed hanging when scrolling over single line input fields with `scroll_beyond_last_line` set to `vertical_scroll_margin`. The change here is to fix the calculations of available lines.

The issue only arises with the setting present because with all overscroll settings and `max_row` being 1 for single-line editors, the calculation would still return the correct value of available lines, which is 1. However, with overscrolling set to `vertical_scroll_margin` and that set to any value greater than 0, the calculation would return that the single-line editor has more than one line, which caused the issues described above (Actually, setting `vertical_scroll_margin` to 1 works for some reason, overscrolls "properly" and does not cause a crash. But I really did not want to investigate this buggy behavior further). 

This PR fixes this by always reporting the number of available lines as the line number value for single line editors, which will (mostly) be 1 (for more context see the discussion in this PR).

Release Notes:

- Fixed an issue where Zed would crash when scrolling over single line input fields and `scroll_beyond_last_line` set to `vertical_scroll_margin`.

MrSubidubi committed:
editor: Fix hang on scrolling single line input fields

MrSubidubi committed:
Also disable scrolling programatically

mocenigo commented:
The only issue I see with such a solution is that we may want to search and replace multi-line snippets. Is this still possible with this fix?

MrSubidubi commented:
> The only issue I see with such a solution is that we may want to search and replace multi-line snippets. Is this still possible with this fix? ...

Yes, you are right - the second commit prevents this from happening. 

While I could revert that change, I think I shouldn't - the current behavior definitely is not intended to be there. 
Single line inputs should not have scrollable content nor be scrollable in any situation. 
What should happen with multiline-contents is that these either are converted into a singe-lined content or something different. 

Also, as you describe yourself, most common keyboard shortcuts for scrolling do not work, as they are also not intended to work. Thus, I would not really consider this a regression of that behavior, but rather a fix-up of a faulty behavior that was in place for a too long time.

---

Whilst I would agree with you that searching for multiline contents should be possible, this discussion goes far beyond the scope of this PR an should happen in a dedicated discussion. This seems off-topic for this specific fix. Supporting multiline searches would possibly require an entire redesign of the UI. Furthermore, once that is supported, the mode of the editor should definitely not be `EditorMode::SingleLine` but rather `EditorMode::AutoHeight`. The latter is unaffected by these changes.

osiewicz reviewed (approved):
Thanks!
(edit: whoops, I did not see the follow-up discussion)

MrSubidubi commented:
> Thanks! (edit: whoops, I did not see the follow-up discussion)

I am not sure whether the is much left to discuss. Issues like https://github.com/zed-industries/zed/issues/28571 for me indicate that the current design is not optimal. Yet I do not really know what can be solved in the scope of this PR to accomodate for that. The solution for the project search would be to use auto height editors instead. But I'd guess that would need a longer discussion as well as a redesign of that interface in a separate PR.

I could add some handling here which would prevent users from pasting multiline contents into single line input fields and replace line breaks with spaces, similiar to how browsers or other apps handle this. With that in place, there would never be content to scroll in single line editors as well as no need for any scrolling then, so the aforementioned issue would be resolved. Personally, I feel that this would make the most sense for single-line input fields.

Curious about what you think would be the right approach here.

Akiyamka commented:
@MrSubidubi, first and foremost, I would like to express my gratitude to you for addressing this issue. However, as the person who created #28571, I believe it is important to highlight that a visual problem and suboptimal design are preferable to the complete absence of such basic functionality as multi-line search and replace.
I agree with you that the issue should be addressed comprehensively; however, it is evident that this will take considerable time. As a user, I would strongly prefer not to lose frequently used functionality during this period due to interface imperfections.
Simply put - this is a case where a poorly functioning feature is better than no feature at all.

MrSubidubi committed:
Revert "Also disable scrolling programatically"

This reverts commit 4b808b2293adf5914a2f773a398640c5dd04c3e3.

MrSubidubi commented:
I reverted the programmatic disabling of scrolling for now given the comment from @Akiyamka . I still think that single-line editors should also only have single-line content and that this should be discussed and addressed further. 
Yet, as discussed above, this will require more changes and more time to properly tackle. So for now I would argue it seems more reasonable to just fix the crash for affected users and deal with the other stuff later.
